helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -294,99 +294,43 @@ private PartitionStateMap computeIntermediatePartitionState(ResourceControllerDa
-        if (!currentStateMap.equals(bestPossibleMap)) {
-          partitionsNeedRecovery.add(partition);
-          isRebalanceNeeded = true;
-        }
-      } else if (rebalanceType.equals(RebalanceType.LOAD_BALANCE)) {
-        // Number of states required by StateModelDefinition are satisfied, but to achieve
-        // BestPossibleState, need load balance
-        partitionsNeedLoadBalance.add(partition);
-        isRebalanceNeeded = true;
-      }
-
-      // Currently at BestPossibleState, no further action necessary
-      if (!isRebalanceNeeded) {
-        Map<String, String> intermediateMap = new HashMap<>(bestPossibleMap);
-        intermediatePartitionStateMap.setState(partition, intermediateMap);
-      }
-    }
-
-    if (!partitionsNeedRecovery.isEmpty()) {
-      LogUtil.logInfo(logger, _eventId, String.format(
-          "Recovery balance needed for %s partitions: %s", resourceName, partitionsNeedRecovery));
-    }
-    if (!partitionsNeedLoadBalance.isEmpty()) {
-      LogUtil.logInfo(logger, _eventId, String.format("Load balance needed for %s partitions: %s",
-          resourceName, partitionsNeedLoadBalance));
-    }
-    if (!partitionsWithErrorStateReplica.isEmpty()) {
-      LogUtil.logInfo(logger, _eventId,
-          String.format("Partition currently has an ERROR replica in %s partitions: %s",
-              resourceName, partitionsWithErrorStateReplica));
-    }
-
-    chargePendingTransition(resource, currentStateOutput, throttleController, cache, preferenceLists, stateModelDef, intermediatePartitionStateMap);
-
-    // Perform recovery balance
-    Set<Partition> recoveryThrottledPartitions =
-        recoveryRebalance(resource, bestPossiblePartitionStateMap, throttleController,
-            intermediatePartitionStateMap, partitionsNeedRecovery, currentStateOutput,
-            cache.getStateModelDef(resource.getStateModelDefRef()).getTopState(), cache);
-
-    // Perform load balance upon checking conditions below
-    Set<Partition> loadbalanceThrottledPartitions;
+    Set<String> messagesForRecovery = new HashSet<>();
+    Set<String> messagesForLoad = new HashSet<>();
+    Set<String> messagesThrottledForRecovery = new HashSet<>();
+    Set<String> messagesThrottledForLoad = new HashSet<>();
     ClusterConfig clusterConfig = cache.getClusterConfig();
 
     // If the threshold (ErrorOrRecovery) is set, then use it, if not, then check if the old
     // threshold (Error) is set. If the old threshold is set, use it. If not, use the default value
     // for the new one. This is for backward-compatibility
     int threshold = 1; // Default threshold for ErrorOrRecoveryPartitionThresholdForLoadBalance
-    int partitionCount = partitionsWithErrorStateReplica.size();
+    // Keep the error count as partition level. This logic only applies to downward state transition determination
+    int partitionCount = (int) currentStateOutput.getCurrentStateMap(resourceName)
```

Comment:
Nit: consider renaming this to `numPartitionsWithErrorReplica`?

Reply:
Sounds good.